contracts/serverless-v2/CommonChainGateways.sol

```diff
@@ -0,0 +1,419 @@
+        address enclaveKey = _pubKeyToAddress(_enclavePubKey);
+        require(
+            gateways[enclaveKey].operator == _msgSender(),
+            "ONLY_GATEWAY_OPERATOR"
+        );
+        _;
+    }
+
+    event GatewayRegistered(
+        bytes enclavePubKey,
+        address indexed enclaveAddress,
+        address indexed operator,
+        uint256[] chainIds
+    );
+
+    event GatewayDeregistered(bytes enclavePubKey);
+
+    event GatewayStakeAdded(
+        bytes enclavePubKey,
+        uint256 addedAmount,
+        uint256 totalAmount
+    );
+
+    event GatewayStakeRemoved(
+        bytes enclavePubKey,
+        uint256 removedAmount,
+        uint256 totalAmount
+    );
+
+    event ChainAddedGlobal(
+        uint256 chainId,
+        address contractAddress,
+        string rpcUrl
+    );
+
+    event ChainRemovedGlobal(
+        uint256 chainId
+    );
+
+    event ChainAdded(
+        bytes enclavePubKey,
+        uint256 chainId
+    );
+
+    event ChainRemoved(
+        bytes enclavePubKey,
+        uint256 chainId
+    );
+
+    error ChainAlreadyExists(uint256 chainId);
+    error ChainNotFound(uint256 chainId);
+
+    function registerGateway(
+        bytes memory _attestation,
+        bytes memory _enclavePubKey,
+        bytes memory _PCR0,
+        bytes memory _PCR1,
+        bytes memory _PCR2,
+        uint256 _timestampInMilliseconds,
+        uint256[] memory _chainIds,
+        bytes memory _signature,
+        uint256 _stakeAmount
+    ) external {
+        // attestation verification
+        _verifyEnclaveKey(
+            _attestation, 
+            IAttestationVerifier.Attestation(_enclavePubKey, _PCR0, _PCR1, _PCR2, _timestampInMilliseconds)
+        );
+
+        // signature check
+        _verifySign(_chainIds, _signature);
+
+        // transfer stake
+        token.safeTransferFrom(_msgSender(), address(this), _stakeAmount);
+        
+        address enclaveKey = _pubKeyToAddress(_enclavePubKey);
+        for (uint256 index = 0; index < _chainIds.length; index++) {
+            require(requestChains[_chainIds[index]].contractAddress != address(0), "UNSUPPORTED_CHAIN");
+        }
+
+        // check missing for validating chainIds array for multiple same chainIds
+
+        gatewayAddresses.push(enclaveKey);
+        
+        gateways[enclaveKey] = Gateway({
+            operator: _msgSender(),
+            chainIds: _chainIds,
+            stakeAmount: _stakeAmount,
+            status: true
+        });
+
+        emit GatewayRegistered(_enclavePubKey, enclaveKey, _msgSender(), _chainIds);
+    }
+
+    function _verifySign(
+        uint256[] memory _chainIds,
+        bytes memory _signature
+    ) internal view {
+        bytes32 digest = keccak256(abi.encodePacked(_chainIds));
+        address signer = digest.recover(_signature);
+
+        _allowOnlyVerified(signer);
+    }
+
+    function getGateway(address _address) public view returns (address, uint256[] memory, uint256, bool) {
```

Comment:
This can be removed as well, since the mapping is being used directly now.

@rajatlko13